packages/authentication/src/services/client.service.ts

```diff
@@ -0,0 +1,55 @@
+import {Request} from '@loopback/rest';
+
+/**
+ * Services for extracting credentials or token from HTTP request.
+ * 
+ * 
+ * Type `C` stands for the type of your credential object.
+ * 
+ * - For local strategy:
+ * 
+ * A typical credential would be: 
+ * {
+ *   username: username, 
+ *   password: password
+ * }
+ *
+ * - For oauth strategy:
+ * 
+ * A typical credential would be:
+ * {
+ *   clientId: string;
+ *   clientSecret: string;
+ *   callbackURL: string;
+ * }
+ * 
+ * It could be read from a local configuration file in the app
+ * 
+ * - For saml strategy:
+ * 
+ * A typical credential would be:
+ * 
+ * {
+ *   path: string;
+ *   issuer: string;
+ *   entryPoint: string;
+ * }
+ * 
+ * It could be read from a local configuration file in the app.
+ */
+export interface ExtractorService<C> {
+  /**
+   * Extract the credentials from the HTTP layer or read them from the local file.
+   * 
+   * For example, extract the basic auth credential `email` and `password` 
+   * from a HTTP request header.
+   * 
+   * Or for third-party access granting, like login facebook account using oauth 2.0, load
+   * the configuration `clientId`, `client`, `scopes`, `callbackURL` from a local file.
+   * @param request The incoming Request
+   */
+  // Here I hardcoded the request type to be HTTP request,
+  // while a generic type R should be used here to support
+  // other clients like SOAP, GRPC
+  extractCredentials(request: Request): Promise<C>;
```

Comment:
What are your thoughts on the following (re-posted from https://github.com/strongloop/loopback-next/issues/2491#issuecomment-469613023)?

I find it problematic to implement transport specific behavior in a service.

How do we envision that such service would be used from a controller? It's important to describe the `login` endpoint via OpenAPI and include specification of input parameters. Are we expecting the controller method to be decorated with `@param` decorators and then pass the entire request to the login service? That would make it too easy to introduce bugs where the controller is describing different parameters than the service extracts from the request.

IMO, extract & invalidate credentials should not be provided by any service, but should be implemented as controller actions leveraging the services described in #2435 under the hood.

```ts
class UserController {
  constructor (
    // inject response, AccessTokenService, TokenTransportStrategy and LoginService
  ) {}

  @post('/login')
  async login(
    @param() email: string,
    @param() password: string,
  ) {
    const user = await this.loginService.verifyCredentials({email, password});
    const token = await this.tokenService.generateAccessToken(user);
    await this.transportStrategy.serializeAccessToken(this.response);
   // uh oh, what is the return value of this method?
   // if the token is returned in response body, how are we going to describe response schema?
  }
    
  @post('/logout')
  async logout(
    @inject('current access token') token: string
  ) {
    await this.tokenService.invalidateAccessToken(token);
  }
}
```

Reply:
@bajtos Yeah that's the twisted part...the login endpoint is verify special, in the [shopping example](https://github.com/strongloop/loopback4-example-shopping/blob/master/src/controllers/user.controller.ts#L125-L154), login doesn't even invokes the authentication action...

In the new design we come up with a new proposal which moves the credential verification back into the auth strategy. The controller function only generates the access token.

Because the auth strategy's responsibility is taking in a request and returning an authenticated user profile, I suggest let developers decide handling the authentication in the auth action OR in the controller:

- Do authentication in strategy:
  - the strategy will extract and validate the user credentials from request
  - it also verifies the credentials by calling `userService.verifyCredentials()`
  - controller function generates the access token
  - pros
    - consistent with endpoints other than `/login`
  - cons
    - lose the convenience provided by REST decorator(e.g. describing parameters using `maxLength`)

- Do authentication in controller:
  - describe the credentials from request using OpenAPI spec
  - action `parseParam` will parse+validate credentials
  - controller function verify credentials(call `userService.verifyCredentials()`) and generates the access token.
  - pros
    - describe the response and request parameters in a clear way.
  - cons
    - repeating the auth strategy logic(extract+verify credentials) using OpenAPI spec+controller code
     - usually the login endpoints is countable and less than 5 for most of applications, treat them specially wouldn't be a huge effort.